pkg/store/proxy.go

```diff
@@ -255,305 +270,209 @@ func (s *ProxyStore) Series(r *storepb.SeriesRequest, srv storepb.Store_SeriesSe
 	}
 	storeMatchers, _ := storepb.PromMatchersToMatchers(matchers...) // Error would be returned by matchesExternalLabels, so skip check.
 
-	g, gctx := errgroup.WithContext(srv.Context())
-
-	// Allow to buffer max 10 series response.
-	// Each might be quite large (multi chunk long series given by sidecar).
-	respSender, respCh := newCancelableRespChannel(gctx, 10)
-	g.Go(func() error {
-		// This go routine is responsible for calling store's Series concurrently. Merged results
-		// are passed to respCh and sent concurrently to client (if buffer of 10 have room).
-		// When this go routine finishes or is canceled, respCh channel is closed.
-
-		var (
-			seriesSet      []storepb.SeriesSet
-			storeDebugMsgs []string
-			r              = &storepb.SeriesRequest{
-				MinTime:                 r.MinTime,
-				MaxTime:                 r.MaxTime,
-				Matchers:                storeMatchers,
-				Aggregates:              r.Aggregates,
-				MaxResolutionWindow:     r.MaxResolutionWindow,
-				SkipChunks:              r.SkipChunks,
-				QueryHints:              r.QueryHints,
-				PartialResponseDisabled: r.PartialResponseDisabled,
-			}
-			wg = &sync.WaitGroup{}
-		)
-
-		defer func() {
-			wg.Wait()
-			close(respCh)
-		}()
-
-		for _, st := range s.stores() {
-			// We might be able to skip the store if its meta information indicates it cannot have series matching our query.
-			if ok, reason := storeMatches(gctx, st, r.MinTime, r.MaxTime, matchers...); !ok {
-				storeDebugMsgs = append(storeDebugMsgs, fmt.Sprintf("store %s filtered out: %v", st, reason))
-				continue
-			}
+	wg := &sync.WaitGroup{}
 
-			storeDebugMsgs = append(storeDebugMsgs, fmt.Sprintf("Store %s queried", st))
+	storeDebugMsgs := []string{}
+	actualRequest := &storepb.SeriesRequest{
+		MinTime:                 r.MinTime,
+		MaxTime:                 r.MaxTime,
+		Matchers:                storeMatchers,
+		Aggregates:              r.Aggregates,
+		MaxResolutionWindow:     r.MaxResolutionWindow,
+		SkipChunks:              r.SkipChunks,
+		QueryHints:              r.QueryHints,
+		PartialResponseDisabled: r.PartialResponseDisabled,
+		PartialResponseStrategy: r.PartialResponseStrategy,
+	}
 
-			// This is used to cancel this stream when one operation takes too long.
-			seriesCtx, closeSeries := context.WithCancel(gctx)
-			seriesCtx = grpc_opentracing.ClientAddContextTags(seriesCtx, opentracing.Tags{
-				"target": st.Addr(),
-			})
-			defer closeSeries()
+	stores := []Client{}
+	for _, st := range s.stores() {
```

Comment:
What's wrong with scheduling all here?

Reply:
Because later on, we do:
```
		storeResponses = make([]struct {
			responses []*storepb.SeriesResponse
		}, len(stores))
```

To avoid reallocating slices. And, IMHO it is more readable.